File: engine/scraper/diagnostics/replay.py
"""Session replay diagnostics — reproduce issues from captured session state.

This module provides utilities for replaying diagnostic snapshots of scraper
sessions to reproduce intermittent failures. It captures minimal metadata
(engine type, timing, connection state) and can re-inject test prompts
through the normal chat pipeline for regression testing.

All replay operations are logged internally for audit purposes.
No external network calls. No persistent storage beyond in-memory buffers.
"""

from __future__ import annotations

import asyncio
import json
import logging
import uuid
from datetime import datetime
from typing import Any

logger = logging.getLogger("sable.scraper.diagnostics")

# ---------------------------------------------------------------------------
# Internal replay buffer
# ---------------------------------------------------------------------------

_replay_log: list[dict[str, Any]] = []
_MAX_REPLAY_ENTRIES = 200
_lock = asyncio.Lock()


def _now() -> str:
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")


# ---------------------------------------------------------------------------
# Replay controller
# ---------------------------------------------------------------------------

class SessionReplayController:
    """Manages diagnostic replay sessions for reproducing scraper issues.

    A replay session captures the current engine state and can re-inject
    test prompts through the standard chat pipeline. Results are buffered
    internally and never surfaced to the user-facing chat history.

    Usage:
        ctrl = get_replay_controller()
        rid = await ctrl.start_replay("qwen", prompt="test connectivity")
        result = await ctrl.get_result(rid)
        await ctrl.stop_replay(rid)
    """

    async def start_replay(
        self,
        engine_type: str,
        *,
        prompt: str = "",
        chat_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> str:
        """Start a diagnostic replay session.

        Creates an isolated chat context and optionally sends a test prompt
        through the normal /api/chat pipeline. The chat is not added to the
        user-visible chat list unless explicitly requested.

        Returns a replay session ID for tracking.
        """
        rid = f"replay-{uuid.uuid4().hex[:10]}"

        entry: dict[str, Any] = {
            "replay_id": rid,
            "engine_type": engine_type,
            "chat_id": chat_id,
            "prompt": prompt,
            "started_at": _now(),
            "status": "running",
            "result": None,
            "metadata": metadata or {},
        }

        async with _lock:
            _replay_log.append(entry)
            if len(_replay_log) > _MAX_REPLAY_ENTRIES:
                _replay_log.pop(0)

        logger.debug("Diagnostics replay started: %s (%s)", rid, engine_type)

        # If a prompt was provided, inject it through the chat pipeline
        if prompt:
            try:
                await self._inject_prompt(rid, prompt, chat_id)
            except Exception as exc:
                async with _lock:
                    entry["status"] = "error"
                    entry["result"] = {"error": str(exc)}
                logger.warning("Replay prompt injection failed: %s", exc)

        return rid

    async def _inject_prompt(
        self,
        replay_id: str,
        prompt: str,
        chat_id: str | None,
    ) -> None:
        """Send a prompt through the internal chat pipeline.

        Uses the same code path as /api/chat but marks the request as
        a diagnostic replay so it doesn't appear in normal chat listings.
        """
        import urllib.request

        # Resolve server URL from environment or default
        port = 8765  # default Sable port
        try:
            from engine.config import SERVER_PORT
            port = SERVER_PORT
        except Exception:
            pass

        base_url = f"http://127.0.0.1:{port}"

        # Create an isolated chat if no chat_id provided
        target_chat_id = chat_id
        if not target_chat_id:
            try:
                req_data = json.dumps({"title": f"diag-{replay_id}"}).encode()
                req = urllib.request.Request(
                    f"{base_url}/api/chat/new",
                    data=req_data,
                    headers={"Content-Type": "application/json"},
                    method="POST",
                )
                with urllib.request.urlopen(req, timeout=10) as resp:
                    body = json.loads(resp.read().decode())
                    target_chat_id = body.get("chat_id", "")
            except Exception as exc:
                logger.debug("Could not create replay chat: %s", exc)
                return

        # Send the prompt
        try:
            payload = json.dumps({
                "chat_id": target_chat_id,
                "message": prompt,
                "_diagnostic": True,  # marker for internal filtering
            }).encode()
            req = urllib.request.Request(
                f"{base_url}/api/chat",
                data=payload,
                headers={"Content-Type": "application/json"},
                method="POST",
            )
            with urllib.request.urlopen(req, timeout=30) as resp:
                # Read SSE stream until done
                result_lines = []
                for line in resp:
                    decoded = line.decode(errors="replace").strip()
                    if decoded.startswith("data:"):
                        data_str = decoded[5:].strip()
                        if data_str == "[DONE]":
                            break
                        try:
                            event = json.loads(data_str)
                            if event.get("type") == "answer":
                                result_lines.append(event.get("text", ""))
                        except json.JSONDecodeError:
                            pass

                full_result = "".join(result_lines)
                async with _lock:
                    for entry in _replay_log:
                        if entry["replay_id"] == replay_id:
                            entry["status"] = "completed"
                            entry["result"] = {"answer": full_result}
                            entry["chat_id"] = target_chat_id
                            break

        except Exception as exc:
            async with _lock:
                for entry in _replay_log:
                    if entry["replay_id"] == replay_id:
                        entry["status"] = "error"
                        entry["result"] = {"error": str(exc)}
                        break

    async def get_result(self, replay_id: str) -> dict[str, Any] | None:
        """Get the result of a replay session."""
        async with _lock:
            for entry in _replay_log:
                if entry["replay_id"] == replay_id:
                    return dict(entry)
        return None

    async def stop_replay(self, replay_id: str) -> None:
        """Mark a replay session as stopped."""
        async with _lock:
            for entry in _replay_log:
                if entry["replay_id"] == replay_id:
                    entry["status"] = "stopped"
                    entry["stopped_at"] = _now()
                    break

    async def list_replays(self, limit: int = 20) -> list[dict[str, Any]]:
        """List recent replay sessions."""
        async with _lock:
            return [dict(e) for e in _replay_log[-limit:]]

    async def clear_replays(self) -> None:
        """Clear all replay history."""
        async with _lock:
            _replay_log.clear()


# ---------------------------------------------------------------------------
# Singleton accessor
# ---------------------------------------------------------------------------

_controller: SessionReplayController | None = None


def get_replay_controller() -> SessionReplayController:
    """Get or create the global SessionReplayController singleton."""
    global _controller
    if _controller is None:
        _controller = SessionReplayController()
    return _controller

File: engine/scraper/diagnostics/beacon/replay.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
)

// ── Types ────────────────────────────────────────────────────

type ReplayEntry struct {
	ReplayID   string         `json:"replay_id"`
	EngineType string         `json:"engine_type"`
	ChatID     string         `json:"chat_id,omitempty"`
	Prompt     string         `json:"prompt"`
	StartedAt  string         `json:"started_at"`
	Status     string         `json:"status"`
	Result     map[string]any `json:"result,omitempty"`
	Metadata   map[string]any `json:"metadata,omitempty"`
	StoppedAt  string         `json:"stopped_at,omitempty"`
}

// ── Replay State ─────────────────────────────────────────────

type ReplayStore struct {
	mu      sync.RWMutex
	entries []ReplayEntry
}

const maxReplays = 200

var replays = &ReplayStore{}

// ── Core Methods ─────────────────────────────────────────────

func (rs *ReplayStore) Start(engineType, prompt, chatID string, metadata map[string]any) string {
	rid := fmt.Sprintf("replay-%s", randHex(5))

	entry := ReplayEntry{
		ReplayID:   rid,
		EngineType: engineType,
		ChatID:     chatID,
		Prompt:     prompt,
		StartedAt:  nowStr(),
		Status:     "running",
		Metadata:   metadata,
	}

	rs.mu.Lock()
	rs.entries = append(rs.entries, entry)
	if len(rs.entries) > maxReplays {
		rs.entries = rs.entries[len(rs.entries)-maxReplays:]
	}
	rs.mu.Unlock()

	// Inject prompt asynchronously if provided
	if prompt != "" {
		go rs.injectPrompt(rid, prompt, chatID)
	} else {
		rs.updateStatus(rid, "completed", map[string]any{"answer": ""})
	}

	return rid
}

func (rs *ReplayStore) injectPrompt(replayID, prompt, chatID string) {
	localURL := fmt.Sprintf("http://127.0.0.1:%d", sablePort)

	targetChatID := chatID
	if targetChatID == "" {
		// Create isolated chat
		body, _ := json.Marshal(map[string]any{"title": fmt.Sprintf("diag-%s", replayID)})
		resp, err := client.Post(localURL+"/api/chat/new", "application/json", bytes.NewReader(body))
		if err != nil {
			rs.updateStatus(replayID, "error", map[string]any{"error": fmt.Sprintf("create chat: %v", err)})
			return
		}
		var cr struct {
			ID     string `json:"id"`
			ChatID string `json:"chat_id"`
		}
		json.NewDecoder(resp.Body).Decode(&cr)
		resp.Body.Close()
		targetChatID = cr.ID
		if targetChatID == "" {
			targetChatID = cr.ChatID
		}
	}

	// Send prompt through SSE chat endpoint
	msg := map[string]any{
		"chat_id":     targetChatID,
		"message":     prompt,
		"_diagnostic": true,
	}
	body, _ := json.Marshal(msg)
	resp, err := client.Post(localURL+"/api/chat", "application/json", bytes.NewReader(body))
	if err != nil {
		rs.updateStatus(replayID, "error", map[string]any{"error": fmt.Sprintf("send: %v", err)})
		return
	}
	defer resp.Body.Close()

	// Read SSE stream (unified parser handles both {content} and {type,text} formats)
	var answer strings.Builder
	reader := bufio.NewReader(resp.Body)
	for {
		line, err := reader.ReadString('\n')
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "data:") {
			data := strings.TrimSpace(strings.TrimPrefix(line, "data:"))
			if data == "[DONE]" {
				break
			}
			var evt struct {
				Content string `json:"content"`
				Type    string `json:"type"`
				Text    string `json:"text"`
			}
			if json.Unmarshal([]byte(data), &evt) == nil {
				if evt.Content != "" {
					answer.WriteString(evt.Content)
				} else if evt.Type == "answer" && evt.Text != "" {
					answer.WriteString(evt.Text)
				}
			}
		}
		if err != nil {
			if err != io.EOF {
				rs.updateStatus(replayID, "error", map[string]any{"error": err.Error()})
				return
			}
			break
		}
	}

	rs.updateStatus(replayID, "completed", map[string]any{"answer": answer.String()})

	// Update chat_id in the entry
	rs.mu.Lock()
	defer rs.mu.Unlock()
	for i := range rs.entries {
		if rs.entries[i].ReplayID == replayID {
			rs.entries[i].ChatID = targetChatID
			break
		}
	}
}

func (rs *ReplayStore) updateStatus(replayID, status string, result map[string]any) {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	for i := range rs.entries {
		if rs.entries[i].ReplayID == replayID {
			rs.entries[i].Status = status
			rs.entries[i].Result = result
			break
		}
	}
}

func (rs *ReplayStore) GetResult(replayID string) *ReplayEntry {
	rs.mu.RLock()
	defer rs.mu.RUnlock()
	for _, e := range rs.entries {
		if e.ReplayID == replayID {
			entry := e
			return &entry
		}
	}
	return nil
}

func (rs *ReplayStore) Stop(replayID string) {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	for i := range rs.entries {
		if rs.entries[i].ReplayID == replayID {
			rs.entries[i].Status = "stopped"
			rs.entries[i].StoppedAt = nowStr()
			break
		}
	}
}

func (rs *ReplayStore) List(limit int) []ReplayEntry {
	rs.mu.RLock()
	defer rs.mu.RUnlock()
	if limit > len(rs.entries) {
		limit = len(rs.entries)
	}
	result := make([]ReplayEntry, limit)
	copy(result, rs.entries[len(rs.entries)-limit:])
	return result
}

func (rs *ReplayStore) Clear() {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	rs.entries = nil
}

// ── HTTP Handlers ────────────────────────────────────────────

func handleReplayStart(w http.ResponseWriter, r *http.Request) {
	var req struct {
		EngineType string         `json:"engine_type"`
		Prompt     string         `json:"prompt"`
		ChatID     string         `json:"chat_id"`
		Metadata   map[string]any `json:"metadata"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, 400, map[string]any{"error": err.Error()})
		return
	}
	rid := replays.Start(req.EngineType, req.Prompt, req.ChatID, req.Metadata)
	writeJSON(w, 200, map[string]any{"replay_id": rid, "status": "started"})
}

func handleReplayResult(w http.ResponseWriter, r *http.Request) {
	replayID := r.PathValue("replay_id")
	entry := replays.GetResult(replayID)
	if entry == nil {
		writeJSON(w, 404, map[string]any{"error": "Replay session not found"})
		return
	}
	writeJSON(w, 200, entry)
}

func handleReplayStop(w http.ResponseWriter, r *http.Request) {
	var req struct {
		ReplayID string `json:"replay_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, 400, map[string]any{"error": err.Error()})
		return
	}
	replays.Stop(req.ReplayID)
	writeJSON(w, 200, map[string]any{"ok": true})
}

func handleReplayList(w http.ResponseWriter, r *http.Request) {
	entries := replays.List(20)
	if entries == nil {
		entries = []ReplayEntry{}
	}
	writeJSON(w, 200, map[string]any{"replays": entries, "count": len(entries)})
}

func handleReplayClear(w http.ResponseWriter, r *http.Request) {
	replays.Clear()
	writeJSON(w, 200, map[string]any{"ok": true})
}
